### ISSUE TITLE
UVR16x2 ?

### ISSUE THREAD

**Issue by Gamerios** [brutella/uvr#1]
**Work or can this be change to work with UVR16x2 ?**

**Comment by brutella** [comment]
This library should work with any CANopen device from Technische Alternative, but I don't have any documentation about the UVR16x2.

**Comment by Gamerios** [comment]
**I tryed not work at UVR16x2**

**Need you the Docs i got from TA /  Technische Alternative?**

**Comment by brutella** [comment]
Please post any documentation you have about the UVR16x2

**Comment by Gamerios** [comment]
**Here all i got from TA by email:**

**https://mega.nz/#!nAsVCBLL!1iucZ3oFRqCDQjJyG8nekHdct-1C4KWSnEN826b1Goo**

For CAN Protocol and UVR16x2


Thank you and i hope you can do something and get UVR16x2 to work. 

If you need anything i can give you can adress index subindex  etc.

**Comment by brutella** [comment]
The same documentation is already available in this repository under the `docs` folder. And there is no mention of the UVR16x2 in the docs.

**Comment by Gamerios** [comment]
**Yes sorry also saw it was TAs fault.  here the newest CAN i got from them for UVR16x2**

**For DL Bus i will get then theay said if they have** 

_https://mega.nz/#!bdECSAqI!XbyimNvUJ_Tl0z96vtxTcC_mtPhJ2IhXpYNj6VzRsgQ_

**Regards**

**Comment by splatch** [comment]
> **Yes sorry also saw it was TAs fault. here the newest CAN i got from them for UVR16x2**
> 
> **For DL Bus i will get then theay said if they have**
> 
> _https://mega.nz/#!bdECSAqI!XbyimNvUJ_Tl0z96vtxTcC_mtPhJ2IhXpYNj6VzRsgQ_
> 
> **Regards**

Hey, would you mind resharing above document? I got very, very short document about uvr16x2 which does not specify index groups making it really useless when using canopen.

**Comment by jdeisenh** [comment]
Wrote some code to get sensor data from uvrx2, based on can/canopen library. 
See https://github.com/jdeisenh/uvrx2
The format is quite different, and there is no documentation that I am aware of. 
What I have is reverse engineered, and might be useful as a basis
![grafana](https://github.com/user-attachments/assets/cf4bdea7-6f84-48b9-8c0a-4f0590b0a6f0)

**Comment by decontamin4t0R** [comment]
Hi,
I also have something like this running, however directly to an mqtt server.
What I am using to write is the following:
```python
b = response[0:1] + int.to_bytes(value, 1, byteorder='little')
msg = b + CrcArc.calc(b).to_bytes(2, 'little')
```

For example pressing and unpressing a button which lives at 0x4F05[01] is:
(triggering horizontal for a window blinds function 2)

``` 
# press
x = get 0x4F05[1] = 0xB1 0x00 0x00 0xFF 0x02
set 0x4F05[1] = 0xB1 0x01 0xB5 0x90 = x[0] + new_value + CrcArc(0xB1 0x01)

# unpress
x = get 0x4F05[1] = 0xB1 0x00 0x00 0xFF 0x02
set 0x4F05[1] = 0xB1 0x00 0x74 0x50 = x[0] + new_value + CrcArc(0xB1 0x00)
```

I hope this was understandable. I have more in my code and mind, but its all not documented, sadly.
For example, the devices expect that the timestamp message contains a flag for DST...

**Comment by jdeisenh** [comment]
Thanks @decontamin4t0R. I tried it, but could not make it work.
Thats for the uvrx2 generation?

**Comment by decontamin4t0R** [comment]
> Thanks @decontamin4t0R. I tried it, but could not make it work. Thats for the uvrx2 generation?

Yes. Depending on the UVR version, it seems like some actions/settings might require opening a separate SDO client and/or authenticating.
